### PR TITLE
Add `spec.files` for Condition and Target of `kind` `yaml`

### PR DESCRIPTION
**Pull request by lemeurherve**
# feat(yaml): implement spec.files for kind: yaml

Follow-up of #490
Ref #435

Allow to specify more than one file as target for `kind: yaml`

Like what have been done for `kind: file`, the possibility to specify multiple files is only possible in the `targets` stage.

## Test

To test this pull request, you can run the following commands:

```shell
cp <to_package_directory>
go test
```

### CODE REVIEW

File: examples/updatecli.generic/yaml-files.yaml [hunk 1]
```diff
@@ -0,0 +1,98 @@
+---
```
Comment (olblak) [inline examples/updatecli.generic/yaml-files.yaml:1]: For information the directory `examples` is deprecated in favor of `e2e/updatecli.d/success/` if the manifest doesn't contain warning or error message, `e2e/updatecli.d/deprecated` if you want to test deprecated message

File: pkg/plugins/resources/yaml/condition.go [hunk 2]
```diff
@@ -18,17 +18,28 @@ func (y *Yaml) Condition(source string) (bool, error) {
...
 
 func (y *Yaml) condition(source string) (bool, error) {
+	var fileContent string
+	var filePath string
 
-	// Test at runtime if a file exist
-	if !y.contentRetriever.FileExists(y.spec.File) {
-		return false, fmt.Errorf("the yaml file %q does not exist", y.spec.File)
+	if len(y.files) > 1 {
+		validationError := fmt.Errorf("Validation error in conditions of type 'yaml': the attributes `spec.files` can't contain more than one element for conditions")
```
Comment (olblak) [inline pkg/plugins/resources/yaml/condition.go:40]: Wouldn't it make sense to allow files in condition as well?
To me it only makes sense to disable it for source

File: pkg/plugins/resources/yaml/main.go [hunk 3]
```diff
@@ -8,29 +8,32 @@ import (
...
 // Spec defines a specification for a "yaml" resource
 // parsed from an updatecli manifest file
 type Spec struct {
-	File    string
-	Key     string
-	Value   string
-	KeyOnly bool // [condition] allow checking for only the existence of a key (not its value)
+	// File contains the file path to take in account
+	File string
+	// Files contains the file path(s) to take in account
```
Comment (olblak) [inline pkg/plugins/resources/yaml/main.go:22]:
```suggestion
	// [target] Files contains the file path(s) to take in account
```

File: pkg/plugins/resources/yaml/main.go [hunk 4]
```diff
@@ -8,29 +8,32 @@ import (
...
+	Files []string
+	// Key is the YAML key to retrieve
```
Comment (olblak) [inline pkg/plugins/resources/yaml/main.go:24]:
```suggestion
	// [source,condition,target] Key is the YAML key to retrieve
```

File: pkg/plugins/resources/yaml/main.go [hunk 5]
```diff
@@ -8,29 +8,32 @@ import (
...
+	Key string
+	// Value is the YAML value to set
```
Comment (olblak) [inline pkg/plugins/resources/yaml/main.go:26]:
```suggestion
	// [source,condition,target] Value is the YAML value to set
```

File: pkg/plugins/resources/yaml/source.go [hunk 6]
```diff
@@ -13,41 +13,51 @@ import (
...
 func (y *Yaml) Source(workingDir string) (string, error) {
 	// By default workingDir is set to local directory
+	var fileContent string
+	var filePath string
 
-	// Test at runtime if a file exist
-	if !y.contentRetriever.FileExists(y.spec.File) {
-		return "", fmt.Errorf("the yaml file %q does not exist", y.spec.File)
+	if len(y.files) > 1 {
+		validationError := fmt.Errorf("Validation error in conditions of type 'yaml': the attributes `spec.files` can't contain more than one element for conditions")
```
Comment (olblak) [inline pkg/plugins/resources/yaml/source.go:20]:
```suggestion
		validationError := fmt.Errorf("Validation error in source of type 'yaml': the attributes `spec.files` can't contain more than one element for conditions")
```

File: pkg/plugins/resources/yaml/source.go [hunk 7]
```diff
@@ -13,41 +13,51 @@ import (
...
+		logrus.Errorf(validationError.Error())
+		return "", validationError
 	}
 
 	if y.spec.Value != "" {
 		logrus.Warnf("Key 'Value' is not used by source YAML")
 	}
 
+	// TODO: warn if the boolean 'KeyOnly' is set?
```
Comment (olblak) [inline pkg/plugins/resources/yaml/source.go:29]: Maybe missing a condition :)

Comment (olblak) [reply pkg/plugins/resources/yaml/source.go:29]: Om other plugins, I create a function `validateTarget` or `validattesource`

File: pkg/plugins/resources/yaml/condition.go (hunk 2 above)
Comment (lemeurherve) [reply pkg/plugins/resources/yaml/condition.go:40]: Let's see what it would look like.
I haven't done this for "type: File", do you think I should too?

**Comment by lemeurherve** [comment]
I didn't managed to reproduce e2e errors locally.
https://github.com/updatecli/updatecli/runs/6985093474?check_suite_focus=true :
image

```
$ ./dist/updatecli diff --config e2e/updatecli.d/success.d/yaml --values e2e/values.yaml
[...]
Run Summary
===========
Pipeline(s) run:
  * Changed:    3
  * Failed:     0
  * Skipped:    0
  * Succeeded:  0
  * Total:      3
```

Any idea?

**Comment by olblak** [comment]
I am spotting several errors

```
⚠ Key 'name', from file 'e2e/updatecli.d/success.d/yamlExample.yaml', was updated from 'Test yaml resource' to 'Test yaml resource change'

ERROR: something went wrong in target "testFilesSourceInput" : "no file was updated"
```

**Comment by olblak** [comment]
The e2e yamlfile that I put to test the PR, are wrong 
```e2e/updatecli.d/success.d/yamlExample*yaml```

**Comment by olblak** [comment]
So the last pipeline that need to work is `updatecli.d/success.d/yaml-files.yaml`

The purpose is to be sure that `make test-e2e` works from the root of the repository

**Comment by dduportal** [comment]
Hello @hervelemeur , what is the status on this PR?

If it should be "ready to test and merge", could you rebase it from the latest upstream `main` branch, fix conflicts? (otherwise moving it back as draft or even closing it)

**Comment by olblak** [comment]
As far as I remember the E2E tests need to be fixed

**Comment by lemeurherve** [comment]
> As far as I remember the E2E tests need to be fixed

Yes, I didn't found out why they're working on my machine and not elsewhere. I'll take any idea ^^

**Comment by dduportal** [comment]
> > As far as I remember the E2E tests need to be fixed
> 
> Yes, I didn't found out why they're working on my machine and not elsewhere. I'll take any idea ^^

Can you rebase + solve conflicts? We'll be able to check after that

**Comment by lemeurherve** [comment]
Another test I've just done:

From the main branch up-to-date, I've added a file `e2e/updatecli.d/success.d/yaml-files.yaml` with only the first source block from https://github.com/updatecli/updatecli/pull/650/files#diff-d0628d5d1105df5256c98fbbf3b0f6196b27de22a53e1d2a9a77675188efc1fcR1-R8:
```
---
title: Show a set of yaml resources as a generic example
sources:
  default:
    kind: yaml
    spec:
      file: examples/values.yaml
      key: "github.user"
```
Here is the "missing file" error I've got:

> #####################################################
> # SHOW A SET OF YAML RESOURCES AS A GENERIC EXAMPLE #
> #####################################################
> 
> 
> SOURCES
> =======
> 
> default
> -------
> ERROR: ✗ the yaml file "examples/values.yaml" does not exist
> Pipeline "Show a set of yaml resources as a generic example" failed
> Skipping due to:
>         sources stage:  "the yaml file \"examples/values.yaml\" does not exist"

**Comment by lemeurherve** [comment]
Note: running this yaml-files.yaml directly is working
```
$ ./dist/updatecli_darwin_amd64_v1/updatecli diff --config e2e/updatecli.d/success.d/yaml-files.yaml

+++++++++++
+ PREPARE +
+++++++++++

Loading Pipeline "e2e/updatecli.d/success.d/yaml-files.yaml"

SCM repository retrieved: 0


++++++++++++
+ PIPELINE +
++++++++++++



#####################################################
# SHOW A SET OF YAML RESOURCES AS A GENERIC EXAMPLE #
#####################################################


SOURCES
=======

default
-------
✔ Value 'olblak' found for key github.user in the yaml file examples/values.yaml

=============================

REPORTS:


- YAML-FILES.YAML:
        Source:
                ✔ [default]  (kind: yaml)

Run Summary
===========
Pipeline(s) run:
  * Changed:    0
  * Failed:     0
  * Skipped:    1
  * Succeeded:  0
  * Total:      1
```

**Comment by olblak** [comment]
Well the recent change that I did in yaml source broke you PR 

```
SOURCES
=======

fromUrl
-------
ERROR: ✗ the yaml file "/home/olblak/Projects/Updatecli/updatecli/https:/raw.githubusercontent.com/updatecli/updatecli/dc90759914718beee17101c693911088cfbf6deb/examples/values.yaml" does not exist

fromExplicitFile
----------------
ERROR: ✗ the yaml file "/home/olblak/Projects/Updatecli/updatecli/file:/examples/values.yaml" does not exist
```

I need to think about a proper fix

**Comment by olblak** [comment]
> Well the recent change that I did in yaml source broke you PR
> 
> ```
> SOURCES
> =======
> 
> fromUrl
> -------
> ERROR: ✗ the yaml file "/home/olblak/Projects/Updatecli/updatecli/https:/raw.githubusercontent.com/updatecli/updatecli/dc90759914718beee17101c693911088cfbf6deb/examples/values.yaml" does not exist
> 
> fromExplicitFile
> ----------------
> ERROR: ✗ the yaml file "/home/olblak/Projects/Updatecli/updatecli/file:/examples/values.yaml" does not exist
> ```
> 
> I need to think about a proper fix

I'll do it in a different PR as I also concern other file resources

**Comment by olblak** [comment]
@lemeurherve I am currently looking at this PR to integrate all the changes that I made

**Comment by hervelemeur** [comment]
I've done a lot this afternoon too, wait a little I'll push what I've already merged and adapted

**Comment by hervelemeur** [comment]
Found the error on autodisvovery tests, will push a fix after the park

**Comment by lemeurherve** [comment]
Green at last! 🎉

File: e2e/updatecli.d/success.d/helmchart.yaml [hunk 8]
```diff
@@ -42,6 +42,7 @@ targets:
     spec:
       name: chart/epinio
       file: Chart.yaml
+      files: []
```
Comment (olblak) [inline e2e/updatecli.d/success.d/helmchart.yaml:45]: Is this line needed?

File: examples/updatecli.generic/yaml-files.yaml [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+title: Show a set of yaml resources as a generic example
```
Comment (olblak) [inline examples/updatecli.generic/yaml-files.yaml:2]: I am wondering if this file should be moved to the e2e directory, I am considering removing the examples directory. 
My motivation is, we don't test them so we don't detect when they are outdated while those in the e2e directory will fail

File: pkg/plugins/resources/yaml/main.go [hunk 10]
```diff
@@ -8,24 +8,24 @@ import (
...
 // parsed from an updatecli manifest file
 type Spec struct {
-	// Specifies the YAML file
-	File string `yaml:",omitempty"`
-	// Specifies the YAML key
-	Key string `yaml:",omitempty"`
-	// Specifies the YAML value
-	Value string `yaml:",omitempty"`
+	// [source,condition,target] File contains the file path to take in account
+	File string
```
Comment (olblak) [inline pkg/plugins/resources/yaml/main.go:21]:
```suggestion
	File string `yaml:",omitempty"`
```

File: pkg/plugins/resources/yaml/main.go [hunk 11]
```diff
@@ -8,24 +8,24 @@ import (
...
+	// TODO: implement files for condition too
+	// [target] Files contains the file path(s) to take in account
+	Files []string
```
Comment (olblak) [inline pkg/plugins/resources/yaml/main.go:24]:
```suggestion
	Files []string `yaml:",omitempty"`
```

File: pkg/plugins/resources/yaml/main.go [hunk 12]
```diff
@@ -8,24 +8,24 @@ import (
...
+	// [source,condition,target] Key is the YAML key to retrieve
+	Key string
```
Comment (olblak) [inline pkg/plugins/resources/yaml/main.go:26]:
```suggestion
	Key string `yaml:",omitempty"`
```

File: pkg/plugins/resources/yaml/main.go (hunk 10 above)
Comment (olblak) [reply pkg/plugins/resources/yaml/main.go:21]: Please not that linting may fail because of the wrong space introduce between File string and `yaml:",omitempty"`

File: pkg/plugins/resources/yaml/main.go [hunk 13]
```diff
@@ -8,24 +8,24 @@ import (
...
+	// [source,condition,target] Value is the YAML value to set
+	Value string
```
Comment (olblak) [inline pkg/plugins/resources/yaml/main.go:28]:
```suggestion 
	Value string `yaml:",omitempty"`
```

**Review by olblak — CHANGES_REQUESTED** [review]
I made some suggestion around the spec tags. They are mandatory  for the manifest upgrade to work

File: examples/updatecli.generic/yaml-files.yaml (hunk 9 above)
Comment (lemeurherve) [reply examples/updatecli.generic/yaml-files.yaml:2]: You've added it here, I don't mind if you want to move it elsewhere ^^

File: e2e/updatecli.d/success.d/helmchart.yaml (hunk 8 above)
Comment (lemeurherve) [reply e2e/updatecli.d/success.d/helmchart.yaml:45]: It was before the addition of the missing jsonschema annotations you suggested below, pushing a fix.

File: examples/updatecli.generic/yaml-files.yaml (hunk 9 above)
Comment (lemeurherve) [reply examples/updatecli.generic/yaml-files.yaml:2]: @olblak  I'd suggest to move them in another PR so we can finally ship this current one, WDYT?

File: pkg/plugins/resources/yaml/source.go [hunk 14]
```diff
@@ -13,43 +13,61 @@ import (
...
 	}
 
-	data := y.currentContent
+	// loop over the only file
+	for theFilePath := range y.files {
+		fileContent = y.files[theFilePath]
+		filePath = theFilePath
+
+		// Merge File path with current workingDir, unless File is an HTTP URL
+		if len(workingDir) > 0 && workingDir != filePath {
```
Comment (lemeurherve) [inline pkg/plugins/resources/yaml/source.go:41]: I'm wondering if this `if len(workingDir) > 0 && workingDir != filePath` tests shouldn't be done directly in joinPathWithWorkingDirectoryPath, that would avoid making these tests in multiple locations (and fixes the places where it's missing?)

File: pkg/plugins/resources/yaml/main.go [hunk 15]
```diff
@@ -8,23 +8,23 @@ import (
...
-)
-
 // Spec defines a specification for a "yaml" resource
 // parsed from an updatecli manifest file
 type Spec struct {
-	// Specifies the YAML file
+	// [source,condition,target] File contains the file path to take in account
 	File string `yaml:",omitempty"`
-	// Specifies the YAML key
+	// TODO: implement files for condition too
```
Comment (lemeurherve) [inline pkg/plugins/resources/yaml/main.go:22]:
```suggestion
```
I prefer to implement multiple files for conditions too in another pull request, as I don't see the immediate use case(s).

**Review by olblak — DISMISSED** [review]

File: pkg/plugins/resources/yaml/source.go (hunk 14 above)
Comment (olblak) [reply pkg/plugins/resources/yaml/source.go:41]: `len(workingDir) > 0` is already tested in `joinPathWithWorkingDirectoryPath`

but I don't understand why `workingDir != filePath` is needed

**Comment by lemeurherve** [comment]
> @olblak enabled auto-merge (squash) 22 seconds ago

Shouldn't we wait a review from @dduportal ?

**Comment by olblak** [comment]
> > @olblak enabled auto-merge (squash) 22 seconds ago
> 
> Shouldn't we wait a review from @dduportal ?

+1 if he has time

**Comment by olblak** [comment]
But unless major issue, I would move forward with the current PR

**Comment by lemeurherve** [comment]
> > > @olblak enabled auto-merge (squash) 22 seconds ago
> > 
> > 
> > Shouldn't we wait a review from @dduportal ?
> 
> +1 if he has time

I'll ask him. This pull request can wait a little bit longer ^^

File: pkg/plugins/resources/yaml/source.go (hunk 14 above)
Comment (lemeurherve) [reply pkg/plugins/resources/yaml/source.go:41]: I've encountered some cases where `workingDir` was set to the same value as `filePath`, and got this as final path without this: `test.yaml/test.yaml`

Comment (lemeurherve) [reply pkg/plugins/resources/yaml/source.go:41]: > `len(workingDir) > 0` is already tested in `joinPathWithWorkingDirectoryPath`

This one is on me, botched cherry-picking

**Comment by lemeurherve** [comment]
@dduportal @olblak WDYT of this PR? Should it be merged? Closed?

**Comment by olblak** [comment]
I would definitely be interested by this one. Last time I checked it was good but that was a while so let me take another look

**Comment by lemeurherve** [comment]
@olblak @dduportal PTAL

File: pkg/plugins/resources/yaml/condition.go [hunk 16]
```diff
@@ -16,18 +16,28 @@ func (y *Yaml) Condition(source string) (bool, error) {
...
 }
 
 func (y *Yaml) condition(source string) (bool, error) {
+	var fileContent string
+	var filePath string
 
-	// Test at runtime if a file exist
-	if !y.contentRetriever.FileExists(y.spec.File) {
-		return false, fmt.Errorf("the yaml file %q does not exist", y.spec.File)
+	if len(y.files) > 1 {
```
Comment (olblak) [inline pkg/plugins/resources/yaml/condition.go:37]: Is there any reason to not check for more than one condition? 
I can see some value here and I don't think it requires a lot of code change

File: pkg/plugins/resources/yaml/main.go [hunk 17]
```diff
@@ -8,23 +8,22 @@ import (
...
-)
-
 // Spec defines a specification for a "yaml" resource
 // parsed from an updatecli manifest file
 type Spec struct {
-	// Specifies the YAML file
+	// [source,condition,target] File contains the file path to take in account
 	File string `yaml:",omitempty"`
-	// Specifies the YAML key
+	// [target] Files contains the file path(s) to take in account
```
Comment (olblak) [inline pkg/plugins/resources/yaml/main.go:22]:
```suggestion
	// [source][condition][target] Files contains the file path(s) to take in account. Source and Condition are limited to one item
```

File: pkg/plugins/resources/yaml/main.go [hunk 18]
```diff
@@ -8,23 +8,22 @@ import (
...
+	Files []string `yaml:",omitempty"`
+	// [source,condition,target] Key is the YAML key to retrieve
 	Key string `yaml:",omitempty"`
-	// Specifies the YAML value
+	// [source,condition,target] Value is the YAML value to set
```
Comment (olblak) [inline pkg/plugins/resources/yaml/main.go:26]:
```suggestion
	// [source,condition,target] Value is the YAML value to set. Default value set to source output for condition and target
```

**Comment by olblak** [comment]
I made a final review with minor suggestion and ready to merge.
I'll wait a day or two before merging

File: pkg/plugins/resources/xml/condition.go [hunk 19]
```diff
@@ -17,6 +17,7 @@ func (x *XML) ConditionFromSCM(source string, scm scm.ScmHandler) (bool, error)
 
 	if scm != nil {
 		x.spec.File = joinPathWithWorkingDirectoryPath(x.spec.File, scm.GetDirectory())
+		logrus.Debugf("Relative path detected: changing to absolute path from SCM: %q", x.spec.File)
```
Comment (dduportal) [inline pkg/plugins/resources/xml/condition.go:20]: This change is not related to this PR: wrong rebase?

File: pkg/plugins/resources/xml/target.go [hunk 20]
```diff
@@ -32,6 +32,7 @@ func (x *XML) TargetFromSCM(source string, scm scm.ScmHandler, dryRun bool) (cha
 
 	if scm != nil {
 		x.spec.File = joinPathWithWorkingDirectoryPath(x.spec.File, scm.GetDirectory())
+		logrus.Debugf("Relative path detected: changing to absolute path from SCM: %q", x.spec.File)
```
Comment (dduportal) [inline pkg/plugins/resources/xml/target.go:35]: This change is not related to this PR: wrong rebase?

File: examples/updatecli.generic/yaml-files.yaml [hunk 21]
```diff
@@ -0,0 +1,100 @@
...
```
Comment (dduportal) [inline examples/updatecli.generic/yaml-files.yaml:2]: I'm late to the party, sorry.

But I thought the `examples/` folder was long deprecated (and removed) because its content is unmaintained.


I don't mind adding an example, for the sake of communication, but an end to end test is required then

File: pkg/plugins/resources/xml/condition.go (hunk 19 above)
Comment (lemeurherve) [reply pkg/plugins/resources/xml/condition.go:20]: I don't think it's from a wrong rebase, the related commit is "chore: add debug logs"

I agree it's not directly related to this PR, I must have added it last september wanting to do good.

I can remove it no problem, but as it's only additional logs, WDYT keeping it?

File: pkg/plugins/resources/xml/target.go (hunk 20 above)
Comment (lemeurherve) [reply pkg/plugins/resources/xml/target.go:35]: Same as above.

File: examples/updatecli.generic/yaml-files.yaml (hunk 21 above)
Comment (lemeurherve) [reply examples/updatecli.generic/yaml-files.yaml:2]: It's been too long ago, I can't remember why it's there 🤷 

I don't mind removing the example.

Comment (lemeurherve) [reply examples/updatecli.generic/yaml-files.yaml:2]: I vaguely remember @olblak saying something about that. Does it ring a bell to you?

Should I just move this file to one of the e2e folders?

File: pkg/plugins/resources/xml/condition.go (hunk 19 above)
Comment (dduportal) [reply pkg/plugins/resources/xml/condition.go:20]: It changes a file unrelated to the PR: I don't mind a PR explaining the intent (since it's done twice) but it can't stay in this one. (same for the other) otherwise the whole PR process (with labels and changelog) would not make sense

File: examples/updatecli.generic/yaml-files.yaml (hunk 21 above)
Comment (lemeurherve) [reply examples/updatecli.generic/yaml-files.yaml:2]: Oh... Check the previous resolved conversation on this file: https://github.com/updatecli/updatecli/pull/650/files#r968050751

This example predates the e2e tests introdution.

File: pkg/plugins/resources/yaml/main.go [hunk 22]
```diff
@@ -52,28 +51,50 @@ func New(spec interface{}) (*Yaml, error) {
...
+	for _, v := range values {
+		uniqueValues[v] = ""
+	}
 
+	return len(values) != len(uniqueValues)
 }
 
 // Validate validates the object and returns an error (with all the failed validation messages) if it is not valid
-func (y *Yaml) Validate() error {
+func (s *Spec) Validate() error {
```
Comment (dduportal) [inline pkg/plugins/resources/yaml/main.go:82]: 👍 good thing to restrict the `Validate()` method to the scope it really validates

File: pkg/plugins/resources/yaml/main.go [hunk 23]
```diff
@@ -52,28 +51,50 @@ func New(spec interface{}) (*Yaml, error) {
...
 	var validationErrors []string
 
 	// Check for all validation
-	if y.spec.File == "" {
-		validationErrors = append(validationErrors, "Invalid spec for yaml resource: 'file' is empty.")
+	if len(s.Files) == 0 && len(s.File) == 0 {
```
Comment (dduportal) [inline pkg/plugins/resources/yaml/main.go:86]:
```suggestion
	if len(s.Files) == 0 && s.File == "" {
```

The compiled code [is the same](https://stackoverflow.com/questions/18594330/what-is-the-best-way-to-test-for-an-empty-string-in-go/54804858#54804858) but in this context, it's clearer to mark that `s.File` is a string (using length is not explicit: it could be a collection)

File: pkg/plugins/resources/yaml/main.go [hunk 24]
```diff
@@ -52,28 +51,50 @@ func New(spec interface{}) (*Yaml, error) {
...
+		validationErrors = append(validationErrors, "Invalid spec for yaml resource: both 'file' and 'files' are empty.")
 	}
-	if y.spec.Key == "" {
+	if s.Key == "" {
 		validationErrors = append(validationErrors, "Invalid spec for yaml resource: 'key' is empty.")
 	}
+	if len(s.Files) > 0 && len(s.File) > 0 {
+		validationErrors = append(validationErrors, "Validation error in target of type 'yaml': the attributes `spec.file` and `spec.files` are mutually exclusive")
+	}
+	if len(s.Files) > 1 && hasDuplicates(s.Files) {
```
Comment (dduportal) [inline pkg/plugins/resources/yaml/main.go:95]: 👍 good one!

File: pkg/plugins/resources/yaml/source.go [hunk 25]
```diff
@@ -12,43 +12,61 @@ import (
...
 
 	valueFound, value, _ := replace(&out, parseKey(y.spec.Key), y.spec.Value, 1)
 
 	if valueFound {
-		logrus.Infof("%s Value '%v' found for key %v in the yaml file %v", result.SUCCESS, value, y.spec.Key, y.spec.File)
+		logrus.Infof("%s Value '%v' found for key %v in the yaml file %v", result.SUCCESS, value, y.spec.Key, filePath)
 		return value, nil
 	}
 
+	// TODO: return result.WARNING? Or an actual error?
```
Comment (dduportal) [inline pkg/plugins/resources/yaml/source.go:65]:
```suggestion
```

File: pkg/plugins/resources/yaml/main.go (hunk 23 above)
Comment (lemeurherve) [reply pkg/plugins/resources/yaml/main.go:86]: Is there something you expect me to do here or is it just a comment on the change?

File: pkg/plugins/resources/xml/condition.go (hunk 19 above)
Comment (lemeurherve) [reply pkg/plugins/resources/xml/condition.go:20]: Moved here: https://github.com/updatecli/updatecli/pull/1119

File: pkg/plugins/resources/yaml/main.go (hunk 23 above)
Comment (dduportal) [reply pkg/plugins/resources/yaml/main.go:86]: no it is a code suggestion to change all the `len(s.File) == 0` occurences that this PR added/updated to `s.File == ""`

**Review by dduportal — APPROVED** [review]

**Review by olblak — APPROVED** [review]

File: pkg/plugins/resources/yaml/main.go [hunk 26]
```diff
@@ -52,28 +51,50 @@ func New(spec interface{}) (*Yaml, error) {
...
-	if y.spec.File == "" {
-		validationErrors = append(validationErrors, "Invalid spec for yaml resource: 'file' is empty.")
+	if len(s.Files) == 0 && s.File == "" {
+		validationErrors = append(validationErrors, "Invalid spec for yaml resource: both 'file' and 'files' are empty.")
 	}
-	if y.spec.Key == "" {
+	if s.Key == "" {
 		validationErrors = append(validationErrors, "Invalid spec for yaml resource: 'key' is empty.")
 	}
+	if len(s.Files) > 0 && s.File == "" {
```
Comment (olblak) [inline pkg/plugins/resources/yaml/main.go:92]:
```suggestion
	if len(s.Files) > 0 && s.File != "" {
```

Comment (lemeurherve) [reply pkg/plugins/resources/yaml/main.go]: Good catch, thanks!
